### PR TITLE
feat(mlx): support native audio training for new omni VLMs

### PR DESCRIPTION
**Pull request by Lyxot**
Adds native audio-input inference and SFT support for the new MLX omni implementations: Nemotron 3 Nano Omni, Gemma 4 Unified, and Qwen3 Omni. Each family is admitted from the first `mlx-vlm` release that contains its complete working audio path; later final releases follow Unsloth's latest-compatible resolver policy instead of requiring a manually maintained ceiling.

Before this, the new families were either refused by the audio gate, lost their required prompt/media ordering, exposed audio modules to full-fine-tuning optimizers, or handed the processor a batch representation that changed the meaning of the audio.

## What changed

- **Nemotron 3 Nano Omni.** Recognizes `sound_encoder`, `sound_projection`, and `sound_clips`; preserves the outer waveform list even when clips have equal lengths; reads the processor-owned 16 kHz sampling rate; keeps sound modules frozen; and normalizes the published mixed-case model type only to an exact case-insensitive mlx-vlm registry key.
- **Gemma 4 Unified.** Qualifies the encoder-free unified implementation separately from legacy Gemma 4, so it keeps its native merge contract and never receives Gemma 4's incompatible placeholder correction. The processor retry removes only the modality-only `audio_kwargs` shield that Gemma 4 forwards to its tokenizer.
- **Qwen3 Omni.** Preserves Qwen's required video/audio/text ordering for scalar counts and structured conversations, delegates final rendering to the native processor without injecting `enable_thinking=False`, honors explicit media-suppression/template options, and freezes the nested Thinker audio tower plus out-of-scope Talker/Code2Wav modules for input-only training.
- **Family qualification.** `gemma4_unified` starts at `mlx-vlm 0.6.5`, Nemotron at `0.5.0`, and Qwen3 Omni at `0.6.7`. Gemma's floor is the first release with its processor/audio-layout repair; Qwen's is the first with both batched-mel and sample-domain-length repairs. Prerelease, post-release, local, malformed, and below-floor versions remain refused, while later compatible final releases remain admitted.

## Validation

The main real-model matrix used `mlx-vlm==0.6.10` and the release environment's `transformers==4.57.6`. Boundary probes additionally ran the exact upstream source tags: Gemma on `v0.6.5` and Qwen on `v0.6.7`.

### Real-dataset 30-step audio SFT

Following the validation shape used in #966, each family trained for 30 LoRA steps on the same ordered speech data. The dataset was `hf-internal-testing/librispeech_asr_dummy` (`clean`, `validation`): despite the repository name, it contains 73 genuine 16 kHz LibriSpeech utterances and transcripts. Each user turn supplied the waveform and a transcription request; the assistant target was the corresponding transcript.

Shared settings were batch size 1, maximum sequence length 512, `torch_randperm` ordering with seed 3407, constant learning rate `1e-4`, gradient checkpointing, standard cross entropy, and LoRA `r=8`, `alpha=16` on Q/V projections in the final language attention layer. Nemotron uses `finetune_last_n_layers=10` only because its final attention block is layer 42; layers 43-51 are MoE/Mamba blocks with no Q/V projections, so its effective adapted scope remains one attention layer.

| family | checkpoint | step 1 loss | step 30 loss | mean loss | mean grad norm | adapter tensors updated | peak memory |
|---|---|---:|---:|---:|---:|---:|---:|
| Gemma 4 Unified | `mlx-community/gemma-4-12B-it-OptiQ-4bit` | 5.86993 | 5.09128 | 5.51637 | 3.53400 | 2/2 | 11.91 GB |
| Qwen3 Omni | `mlx-community/Qwen3-Omni-30B-A3B-Instruct-4bit` | 4.67775 | 4.05842 | 4.25655 | 3.91832 | 4/4 | 31.83 GB |
| Nemotron 3 Nano Omni | `nvidia/Nemotron-3-Nano-Omni-30B-A3B-Reasoning-BF16` (runtime 4-bit affine) | 3.69480 | 3.66736 | 3.55826 | 0.95607 | 4/4 | 27.24 GB |

All 90 losses and gradient norms were finite. Every targeted adapter tensor changed, while every audio encoder/output module remained absent from the trainable parameter set. The exact 30 utterance IDs and order matched across all three runs. Raw per-step losses are not expected to decrease monotonically because each step uses a different utterance and transcript.

### Inference and modality controls

| family | result |
|---|---|
| Nemotron 3 Nano Omni | Audio-conditioned generation passed; exactly transcribed `The quick brown fox jumps over the lazy dog.` |
| Gemma 4 Unified | Audio-conditioned generation passed; exactly transcribed `The quick brown fox jumps over the lazy dog.` on the exact `v0.6.5` floor. |
| Qwen3 Omni | Audio-conditioned generation passed; exactly transcribed `The quick brown fox jumps over the lazy dog.` and preserved the text-only control on the exact `v0.6.7` floor. |

The controlled Nemotron and Qwen runs changed only waveform content while preserving token IDs and audio length, and produced different losses. This proves the training path remained audio-conditioned rather than silently optimizing text alone. On `v0.6.7`, Qwen also completed an additional one-step floor probe with losses `5.83892` and `5.32614` for the controlled waveforms, finite training loss `5.75023`, gradient norm `3.52930`, and all four adapter tensors updated while audio modules remained frozen.

The immediate predecessor explains Qwen's floor: on the same 39,946-sample speech input, `v0.6.6` interpreted the sample-domain mask as mel frames and derived 5,193 audio placeholders; `v0.6.7` normalized the length and derived 33.

### Why there is no CUDA comparison

No CUDA parity result is claimed. The available CUDA host cannot load the 66 GB Nemotron BF16 checkpoint, and the cached Gemma 4 Unified and Qwen3 Omni validation artifacts are MLX conversions rather than same-weight CUDA checkpoints. Without the same weights, rows, and supervised positions, a loss comparison would not be meaningful.

## Known limits

- Qwen3 Omni support is audio-input conditioning of the Thinker/language adapters. It does not train the Talker or Code2Wav audio-output path.
- `_AUDIO_QUALIFIED_FAMILIES` gates audio-bearing training rows; it is not a generic model-load or Studio-inference preflight. Studio's active stale environment was synchronized separately to the latest-compatible `mlx-vlm 0.6.10`.
- The Studio-level renderer lookup for Nemotron's published mixed-case model type belongs to the separate `unsloth` repository change; this PR owns the zoo runtime and training boundary.

## Test coverage

- `233 passed` in `tests/test_mlx_vlm_label_masks.py`
- `55 passed` in `tests/test_mlx_save_export_regressions.py`
- Python compilation and `git diff --check`
- Every changed behavior was exercised test-first. Reverting either corrected family floor makes its boundary proof fail; reintroducing a `0.6.10` ceiling makes both the public gate and future-version proofs fail.
- The complete branch passed the mandatory review gate with no remaining in-scope findings.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/loader.py [hunk 1]
```diff
@@ -5203,6 +5203,39 @@ def _first_media_user_message_index(messages):
     return -1
 
 
+def _qwen3_omni_media_counts(content):
+    counts = [0, 0, 0]
+    for item in content if isinstance(content, list) else ():
+        if not isinstance(item, dict): continue
+        kind = str(item.get("type", "")).lower()
+        index = 0 if kind == "image" or "image" in item or "image_url" in item else 1 if kind == "audio" or "audio" in item or "audio_url" in item else 2 if kind == "video" or "video" in item else None
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py]: Checked against the exact cached Qwen3 Omni processor and the patched public helper on qualified mlx-vlm 0.6.10. Qwen native rendering of `{"type":"input_audio","input_audio":"a.wav"}` produces zero `<|audio_pad|>` placeholders; the same is true for `input_image`/`input_image`. The current helper intentionally preserves the alias object and adds one canonical bare marker, producing exactly one placeholder for the separately counted tensor. If `_qwen3_omni_media_counts` counted that native-unrenderable alias as already present, it would suppress the canonical marker and produce zero placeholders—the opposite of the requested correction. Training message normalization also canonicalizes audio alias types before this seam when the embedded payload owns the clip. Therefore this is not a defect and no code change is appropriate.

File: unsloth_zoo/mlx/loader.py [hunk 2]
```diff
@@ -5453,6 +5507,44 @@ def patched_apply_chat_template(
...
+            and not isinstance(prompt, (dict, list))
+            and num_audios > 0
+        ):
+            # Qwen requires media before text; mlx-vlm renders audio last.
+            message = prompt_utils.get_message_json(
+                model_type,
+                str(prompt),
+                num_images=num_images,
+                num_audios=num_audios,
+                **kwargs,
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py:5597]: Checked the admitted upstream floor and the exercised release. mlx-vlm v0.6.7 and v0.6.10 both guard counted image/audio construction with `role == "user"` inside `MessageFormatter`; assistant and system roles do not materialize those entries. Through the patched public helper on 0.6.10, scalar assistant/system calls with `num_images=1, num_audios=1` return text-only messages and the exact cached Qwen processor renders zero image/audio placeholders. The later normalizer likewise adds media only to user-like roles. The reported formatter-generated non-user media does not reproduce on any qualified dependency version, so forcing redundant skip flags would not fix shipped behavior and no code change is warranted.

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py]: The new premise is contradicted by the exact cached Qwen3 Omni processor. Native rendering of both `{"type":"input_audio","audio":"a.wav"}` and key-only `{"audio":"a.wav"}` produces exactly one `<|audio_pad|>` placeholder; the same key-aware behavior applies to image media. These are precisely the top-level shapes mirrored by `_qwen3_omni_media_counts`. Removing those key checks would make the helper add duplicate canonical markers. The non-rendered shape discussed in the previous thread was different: `{"type":"input_audio","input_audio":"a.wav"}`. That spelling is intentionally not counted, so the helper adds the required canonical marker. No code change was made for this incorrect finding.

**Comment by danielhanchen** [comment]
Ran this across a 16 environment matrix on Linux plus real Apple Silicon CI, and pushed four fixes to the branch. The feature holds up well; the ceiling removal in particular is more clearly correct than the description claims.

### The ceiling removal fixes a currently broken state

I built isolated environments for mlx-vlm 0.4.4, 0.5.0, 0.6.2, 0.6.4, 0.6.5, 0.6.7, 0.6.10 and 0.6.12, each in two flavours: the shipped Apple Silicon resolution (`UV_OVERRIDE=overrides-darwin-arm64.txt` plus `constraints.txt` pinning `transformers==5.5.0`, copied verbatim from the unsloth repo) and mlx-vlm's own preferred resolve.

Two results worth recording:

1. `main`'s comment that 0.6.5+ "cannot be installed here" is not accurate. Under the shipped override it installs fine, and a fresh install with mlx-vlm unpinned resolves to `mlx-vlm==0.6.12` with `transformers==5.5.0`.
2. On that exact environment, with real processors from `mlx-community/gemma-3n-E2B-it-4bit` and `gemma-4-e2b-it-4bit`:

```
main:  gemma3n -> BLOCKED      this PR:  gemma3n -> ALLOWED
       gemma4  -> BLOCKED                gemma4  -> ALLOWED
```

So `main` refuses audio training for four families on the environment its own installer produces, and this PR restores it.

Across all 16 environments there is no cell where `main` admits and this PR refuses, so no backwards compatibility break. On 0.4.4 both admit gemma3n/phi4mm/minicpmo and both refuse the new families.

The un-capping is safe for a reason beyond measurement: `Gemma3nProcessor` and `Gemma4Processor` and their audio feature extractors load from `transformers.models.*`, and transformers is pinned at 5.5.0 regardless of mlx-vlm, so the placeholder expansion contract the gate protects does not vary with the mlx-vlm version. `sha256` of gemma3n and minicpmo `processing_*.py` is byte identical from 0.4.4 through 0.6.12.

Apple Silicon CI, one pytest process per file as `consolidated-tests-ci.yml` does: `test_mlx_vlm_label_masks.py` 233 passed, `test_mlx_save_export_regressions.py` 55 passed, `test_mlx_shape_guard.py` 32 passed. A real mixed case Nemotron `model_type` reaches the renderer on this branch and does not on `main`, which is the fix working on hardware.

### Fix 1: Nemotron's floor should be 0.6.10, not 0.5.0

Below 0.6.10, mlx-vlm's `sanitize_audio_weights` transposes `sound_encoder.encoder.*` convs unconditionally, so an already MLX major weight double transposes and the checkpoint cannot load. Measured on the real function:

```
0.5.0 / 0.6.4 / 0.6.7    (128, 3, 3, 1) -> (128, 3, 1, 3)   corrupted
0.6.10 / 0.6.12          (128, 3, 3, 1) -> (128, 3, 3, 1)   left alone
```

This is the same bug class as gemma4's #1498/#1523, in a second family, and `_ensure_audio_conv_sanitize` does not reach it: that repair keys on the `subsample_conv_projection` and `depthwise_conv1d.weight` markers in the sanitize source, while Nemotron's `Model.sanitize` delegates to a module level function carrying neither. 0.6.10 is also the release your own real model matrix ran on, so this does not narrow anything you validated.

### Fix 2: `freeze_audio_modules` stopped at the first owner

The `break` exited the owner loop after one match, so a model carrying both a top level `audio_tower` and a distinct `thinker.audio_tower` left the nested one trainable, which is exactly the parameter set the call exists to keep out of the optimizer. Now walks every owner, deduped by identity so a shared module is not frozen or reported twice.

### Fix 3: `KeyError` on a turn with no content

`_normalize_qwen3_omni_counted_message({"role": "user"}, 1, 0, {})` raised `KeyError('content')`, where every neighbouring guard returns the message untouched. I traced the call paths and could not reach it in practice, so this is defensive only, but the bare subscript is inconsistent with the rest of the helper.

### Fix 4: `gemma4_unified` transformers floor

`_AUDIO_MIN_TRANSFORMERS` covered `gemma4` but not `gemma4_unified`, though they share the audio path and `gemma4_unified` is gated at an mlx-vlm release implying `transformers>=5.14` while the Apple Silicon override holds transformers at 5.5.0.

Tests: 233 to 237. The new tests fail on the pre-fix code.

### Left for you rather than changed here

These are design calls on your feature, so I did not touch them:

- `_prepare_vlm_template_messages` normalizes ordering only for the first user turn; later media bearing turns keep `[text, audio]`.
- `_qwen3_omni_media_counts` does not recognise the `input_audio`, `input_image`, `input_video` or `video_url` aliases.
- `_NON_USER_ROLES` does not exclude `tool`, so conversation level media can attach to tool output.
- The four bucket partition reorders caller supplied interleaving: `[audioA, textA, audioB, textB]` becomes `[audioA, audioB, textA, textB]`.
- `admits()` is now unbounded forward, so an unreleased mlx-vlm is admitted rather than refused. 0.6.12 is fine (verified), but the "refuse the unproven" property is gone. A warning above the last probed release would keep the open floor and the property.

I deliberately did not add `transformers>=5.14` floors for `nemotron_h_nano_omni` and `qwen3_omni_moe`. That would refuse both families on the shipped 5.5.0 environment and disable what you validated as working.

File: unsloth_zoo/mlx/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/loader.py]: Checked this against the model's own chat template rather than reasoning about it, and the premise does not hold: those aliases do render.

`chat_template.jinja` from `mlx-community/Qwen3-Omni-30B-A3B-Instruct-4bit` uses exactly the predicate the counter uses:

```jinja
{%- if content.type == 'image' or 'image' in content or 'image_url' in content %}
    {{- "<|vision_start|><|image_pad|><|vision_end|>" }}
{%- elif content.type == 'audio' or 'audio' in content or 'audio_url' in content %}
    {{- "<|audio_start|><|audio_pad|><|audio_end|>" }}
{%- elif content.type == 'video' or 'video' in content %}
    {{- "<|vision_start|><|video_pad|><|vision_end|>" }}
```

Rendering each shape through that template:

```
canonical audio      -> <|audio_pad|>
input_audio alias    -> <|audio_pad|>
key-only {audio:...} -> <|audio_pad|>
audio_url            -> <|audio_pad|>
input_image alias    -> <|image_pad|>
```

So `{"type": "input_audio", "audio": ...}` and a key-only `{"audio": ...}` both produce a placeholder, the deficit reduction is correct, and there are not fewer placeholders than tensors. The counter matches any `audio`/`image` key precisely because the template does; they are the same test by construction.

One real thing that check did surface, separate from this item: `video_url` renders nothing at all, since the video branch has no `video_url` clause. That is an upstream template limitation rather than something this PR introduces, and I have noted it on the video_url thread.

**Comment by danielhanchen** [comment]
Correction to something I claimed earlier in this thread, plus the evidence behind the latest round.

I checked the model's own `chat_template.jinja` (from `mlx-community/Qwen3-Omni-30B-A3B-Instruct-4bit`) rather than reasoning about what it probably does, and it settles two items in opposite directions.

**The alias-counting concern does not hold.** The template uses exactly the predicate `_qwen3_omni_media_counts` uses:

```jinja
{%- if content.type == 'image' or 'image' in content or 'image_url' in content %}
    {{- "<|vision_start|><|image_pad|><|vision_end|>" }}
{%- elif content.type == 'audio' or 'audio' in content or 'audio_url' in content %}
    {{- "<|audio_start|><|audio_pad|><|audio_end|>" }}
{%- elif content.type == 'video' or 'video' in content %}
    {{- "<|vision_start|><|video_pad|><|vision_end|>" }}
```

Rendering each shape through it:

```
canonical audio       -> <|audio_pad|>
input_audio alias     -> <|audio_pad|>
key-only {audio: ...} -> <|audio_pad|>
audio_url             -> <|audio_pad|>
input_image alias     -> <|image_pad|>
video_url             -> NONE
```

So the non-canonical audio shapes do render, the deficit reduction is right, and there are no missing placeholders. The counter matches any `audio`/`image` key precisely because the template does.

**My own video_url claim was wrong.** I said grouping `video_url` with video fixed an ordering bug producing `audio, video_url, text`. It does not: `video_url` renders nothing at all, since the video branch has no `video_url` clause, so its position cannot change the rendered prompt. The two counters did genuinely disagree, `_qwen3_omni_media_counts` returning `(0, 0, 0)` where `_structured_multimodal_counts` returns `(0, 0, 1)`, so I kept the grouping change for that consistency and corrected the comment and test docstring in `1dbc855f` to say so rather than claiming an ordering fix.

That check did surface one real gap, which predates this PR: `video_url` is silently dropped by the template while `_structured_multimodal_counts` counts it as a video, so a caller passing `video_url` alongside a video tensor gets no placeholder. Canonicalizing the alias would be a behaviour change beyond this PR's scope, so I have left it and am flagging it instead.
